### ISSUE TITLE
Not catching {...props.children} causes broken es5 file.

### ISSUE THREAD

**Issue by geoguide** [babel/babel#8456]
## Bug Report

**Current Behavior**
When I am running a watch, and I have a `{...props.children}` in my code, it compiles and runs, however when I compile for production, babel leaves `{...props.children}` in the file.

**Input Code**

```js
{...props.children}
```

**Expected behavior/code**
Babel should fail to compile in this case, or convert the spread operator.

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  "presets": ["react", "es2015"],
  "plugins": ["transform-es2015-destructuring", "transform-object-rest-spread"]
}

```

**Environment**
- Babel version(s): [e.g. v6.0.0, v7.0.0-beta.34]
- Node/npm version: Node v10.1.0/npm 5.6.0
- OS: OSX 10.13.4
- Monorepo no
- How you are using Babel: loader

**Possible Solution**
Babel should throw an error on `{...props.children}` instead of succeeding and compiling a broken es2015 file.

**Additional context/Screenshots**

**Comment by loganfsmyth** [comment]
A more complete example would be a big help. It's hard to investigate without a reproducible example, and it's not clear why something like this would only happen in production builds, or what about your build process might be triggering that behavior.

**Comment by geoguide** [comment]
Adding in the original post. It looks like babel just doesn't recognize that `...props.children` is not a valid variable.

**Comment by loganfsmyth** [comment]
I think in this context, you're just supposed to do `{props.children}` without the `...`, though I'm not 100% sure here. This potentially seems related to https://github.com/babel/babel/issues/5492

**Comment by geoguide** [comment]
You're correct, the syntax I used is wrong but it should have thrown an error as `...props.children` is not a valid variable. The issue is that it didn't fail during transpilation and so was deployed when it should have failed to transpile.

**Comment by nicolo-ribaudo** [comment]
Yeah, the parser should throw an error.

**Comment by nicolo-ribaudo** [comment]
I checked the JSX spec and `{...expression}` is a valid element child:

```
JSXChild :
  JSXText
  JSXElement
  { JSXChildExpression_opt }

JSXChildExpression :
  AssignmentExpression
  ... AssignmentExpression
```

It should be supported by the jsx transformer.

**Comment by kevintab95** [comment]
Has this been fixed? 
I get a parsing error when I tried the above code using babylon:
https://astexplorer.net/#/gist/9f9a0266b3bbcfed1520b7c7a11ff8b6/4f429188dfac17d0b4979d5291307407ce087187

**Comment by kevintab95** [comment]
Hi @nicolo-ribaudo --
> I checked the JSX spec and `{...expression}` is a valid element child:
> 
> ```
> JSXChild :
>   JSXText
>   JSXElement
>   { JSXChildExpression_opt }
> 
> JSXChildExpression :
>   AssignmentExpression
>   ... AssignmentExpression
> ```
> It should be supported by the jsx transformer.

I think the spread operator is being used incorrectly. I initialised an array `a` and experimented with the spread operator:

 

input | output
------------ | -------------
`...a` | Uncaught SyntaxError: Unexpected token ...
 `{...a}`| `{0: 1, 1: 2, 2: 3}` (object)
`[...a]`|`[1, 2, 3]` (array)

I observed the same behaviour when using spread operator within JSX. It parses it as expected when we use the spread operator like so:

```
const testDiv = (props) => {
  	var a = [1,2,3];
	return (
      <TestDiv>
      {[...a]}
      </TestDiv>
    );
};
```
see: https://astexplorer.net/#/gist/f6a74ccabda32efbd2392fd4d722115d/c2e26c6d027ceb49a7e359996b986defb88037c3

Am I missing something? Thanks!

**Comment by nicolo-ribaudo** [comment]
You are using an old Babylon version: you should use `babylon7` (which is very old anyway, but astexplorer hasn't been updated to `@babel/parser` yet).
